### ISSUE TITLE
Add Position, Scale, Rotate (Transform) per Clip

### ISSUE THREAD

**Issue by backhausmartin** [PixarAnimationStudios/OpenTimelineIO#253]
Hello, I'm new to OTIO, I hope a can contribute in some way, maybe later also with code.

I need to parse the Position, Scale and Rotation Data on a per-clip basis from an Offline NLE project (in my case Adobe Premiere Pro 2018) - and re-use this information in an Online/Mastering/Comp project (currently Foundry Nuke Studio or just normal Nuke).

I already managed to parse the Premiere Pro Data from an FCP XML via Python xml.etree and figure out how the clipitem.filter.effect.parameter scale, rotation and center tags/ information work in general.
There are five `<parameter authoringApp="PremierePro">` tags `Scale, Rotation, Center, Anchor Point, Anti-flicker Filter` in the XML below the `xmeml.sequence.media.video.track.clipitem.filter.effect` tree. 

Now I wonder if I can use OTIO for this and maybe built it into it - instead of writing a custom script.

How mature is the integration of this transform data to OTIO as of today? I recently checked the code and I think there are only effect placeholders in OTIO? It looks that it is planned, see here, "Audio/Video Effects": https://github.com/PixarAnimationStudios/OpenTimelineIO/wiki/Feature-Matrix

Some Background:
I do an exhibition film project with multiple projectors in a 360° setup. I have a base canvas/box size with a resolution of more than 32.000px x 1.800px (which will be split/cropped to single projectors later). Multiple clips need to be positioned side by side, at the same time in the edit, using the x-position. It's a 10 min linear film.
My goal is to go from the Quarter-Res offline edit project to the Full-Res Online project (semi-)automatically.

If this is interesting I can provide a FCP XML sample data and also do some testing going from Premiere to Nuke via OTIO. But I understand that this is a special case, going away from standard 2K or 4K resolution.

First step: Basic x,y position should be parsed from FCP XML and AAF and saved in OTIO format. Result: A basic reframing of camera footage will be preserved from Editorial to Online/DI.

**Comment by jminor** [comment]
Hi @backhausmartin, that sounds like a great use case for OTIO. You're right that our support for handling effects in OTIO is not complete yet. If you export your timeline from Premiere as XML and then run it through `otiocat` or `otioconvert` you can see what, if any, of that comes through at the moment. 

For many of the adapters, we try to capture information that OTIO doesn't understand as blind metadata so that it can be carried through a pipeline even if we don't represent it in the OTIO data model. For example, we do this with CDL (color correction) information from EDLs. I'm not sure about the status of that for the XML adapter, but @bashesenaxis should know more about that.

Yes, if you can provide a sample XML file that holds the data you are interested in that can help us get right to the details.

**Comment by bashesenaxis** [comment]
The XML adapter will mainly follow the OTIO feature set. So far I have used the metadata feature as little as possible. Because the xml format is so feature rich, it was hard to decide what to include in the metadata and what to leave out. Perhaps it is an idea to keep a suggested list of features to implement as metadata? That way we can aim for some sort of feature parity between adapters, even though we haven`t fully defined the implementation details in OTIO yet? Otherwise I feel there is a risk of widely diverging feature-sets between adapters.
I am happy to look into getting the transform data represented as metadata.

**Comment by ssteinbach** [comment]
@apetrynet Did your PR #264 address the basic need cited by @backhausmartin?
"First step: Basic x,y position should be parsed from FCP XML and AAF and saved in OTIO format. Result: A basic reframing of camera footage will be preserved from Editorial to Online/DI."

Just trying to do some house keeping on the open issues. Thanks!

**Comment by apetrynet** [comment]
@ssteinbach No, unfortunately. I just focused on getting the adapter to accept Hiero produced XML's.

**Comment by backhausmartin** [comment]
Thanks for your efforts and analysis of the topic.

Is it already decided how OTIO will save this data in its internal schema?
And do you know, if all / most FCP XML representations of the data is standardised?

1. Position/Reframing (usually there is a Center/Translate and Center Offset / Anchor Point)
2. Scale
3. Crop (see some research below)
4. Main Project resolution / canvas size

Adobe Premiere Pro, Foundry Nuke and FFmpeg for example represent the data in (slightly) different ways. I guess for basic effects OTIO should use one streamlined format?

In my opinion, especially for "standard" 2K/HD and 4K/UHD live-action feature films shot on Digital Cinema Cameras, it is crucial to preserve this data into the OTIO format.

For example some research, about Crops:

- A Premiere Pro and After Effects Crop uses Distance from Left, Right, Top & Bottom

- Nuke calculates a Crop from left and bottom edge of original image. https://learn.foundry.com/nuke/content/reference_guide/transform_nodes/crop.html?Highlight=crop

- FFmpeg calculates a Crop from left and top edge of output image.
https://ffmpeg.org/ffmpeg-filters.html#crop

It is also necessary to save the main image resolution (canvas size) of a timeline as the FCP XML (Premiere) uses relative positions.

Attached going back to the above mentioned basic need: 
A Screenshot of a working script including debug info about variable data, where I am figuring out how the translate data is represented in FCP XML. It's a single 3200x1800 Arri Alexa clip in a 1998x1080 timeline that is not Fit to the master resolution. 

Because of the relationship between Center/Translate and CenterOffset/Anchor Point it's a little bit more complicated...
I probably should simplify it first, using only one 50px image in a 100px timeline.

![image](https://user-images.githubusercontent.com/10419280/42772614-41c3a1ec-892b-11e8-9a7c-fa82c1e82ad9.png)

**Comment by backhausmartin** [comment]
Here's a direct follow up, I made it simpler and cleaned it up a bit (and the last lines of code are wrong).

I did a test using a 300x100px timeline in Premiere and scaling down and repositioning the 3200x1800 Arri Alexa file to 128x72. It is moved 10px down and 10px to the right from the top left corner.

Sequence Setting: 300x100px
1 single Clip: Arri Alexa 3200x1600px scaled down to 0.04

The effect settings in Premiere Pro (I set the Anchor Point to 0, 0):
![image](https://user-images.githubusercontent.com/10419280/42779028-622b78b0-893f-11e8-8724-df299742dfd1.png)

Resulting Image (you can see the anchor point in top left corner of the image, which is the center in relation to the sequence coordinates):
![image](https://user-images.githubusercontent.com/10419280/42781987-de3182c6-8947-11e8-9fe5-45d9ba4de778.png)

XML output:

[timeline_300x200.zip](https://github.com/PixarAnimationStudios/OpenTimelineIO/files/2199124/timeline_300x200.zip)

Key Section of XML output:

```
<filter>
    <effect>
        <name>Basic Motion</name>
        <effectid>basic</effectid>
        <effectcategory>motion</effectcategory>
        <effecttype>motion</effecttype>
        <mediatype>video</mediatype>
        <pproBypass>false</pproBypass>
        <parameter authoringApp="PremierePro">
            <parameterid>scale</parameterid>
            <name>Scale</name>
            <valuemin>0</valuemin>
            <valuemax>1000</valuemax>
            <value>4</value>
        </parameter>
        <parameter authoringApp="PremierePro">
            <parameterid>rotation</parameterid>
            <name>Rotation</name>
            <valuemin>-8640</valuemin>
            <valuemax>8640</valuemax>
            <value>0</value>
        </parameter>
        <parameter authoringApp="PremierePro">
            <parameterid>center</parameterid>
            <name>Center</name>
            <value>
                <horiz>0.45625</horiz>
                <vert>0.477778</vert>
            </value>
        </parameter>
        <parameter authoringApp="PremierePro">
            <parameterid>centerOffset</parameterid>
            <name>Anchor Point</name>
            <value>
                <horiz>-0.5</horiz>
                <vert>-0.5</vert>
            </value>
        </parameter>
        <parameter authoringApp="PremierePro">
            <parameterid>antiflicker</parameterid>
            <name>Anti-flicker Filter</name>
            <valuemin>0.0</valuemin>
            <valuemax>1.0</valuemax>
            <value>0</value>
        </parameter>
    </effect>
</filter>

```

My Test Code and the output/variable vata:

![image](https://user-images.githubusercontent.com/10419280/42786395-62d82666-8956-11e8-8033-199afadfeed1.png)

https://gist.github.com/backhausmartin/b35fdfa926a717edc823df72c702b069

I tested it with three different Premiere timelines.

Final Note: This Apple document mentions the effect "Center" parameter, but Premiere adds its own parameters (Anchor Point)...
http://mirror.informatimago.com/next/developer.apple.com/documentation/AppleApplications/Conceptual/FinalCutPro_XML/FinalCutPro_XML.pdf

**Comment by jminor** [comment]
Thanks for doing this research. We haven't gotten into the detail of this part of the schema design yet. Our strategy for this sort of thing has been to research and compare how the other formats and tools handle a feature like this and then look for a simple common ground that covers the essential aspects. We have tried to use straight-forward terminology for things and aim for real world use cases.

My only thought on this is that we should try to use normalized coordinates so that relinking a clip to different media, which happens to be a different resolution, wouldn't require updating all of the positional information on an effect.
